src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -1366,6 +1367,61 @@ bool FancyZones::ShouldProcessSnapHotkey(DWORD vkCode) noexcept
     return false;
 }
 
+void FancyZones::ApplyQuickLayout(int vKey) noexcept
+{
+    std::unique_lock writeLock(m_lock);
+
+    int key = vKey - '0';
+
+    HMONITOR monitor;
+    if (m_settings->GetSettings()->spanZonesAcrossMonitors)
+    {
+        monitor = NULL;
+    }
+    else
+    {
+        POINT cursorPoint;
+        if (!GetCursorPos(&cursorPoint))
+        {
+            return;
+        }
+
+        monitor = MonitorFromPoint(cursorPoint, MONITOR_DEFAULTTOPRIMARY);
+    }
+
+    std::wstring uuid;
+    for (auto [zoneUuid, hotkey] : FancyZonesDataInstance().GetLayoutQuickKeys())
+    {
+        if (hotkey == key)
+        {
+            uuid = zoneUuid;
+        }
+    }
+
+    auto workArea = m_workAreaHandler.GetWorkArea(m_currentDesktopId, monitor);
+
+    // Find a custom zone set with this uuid and apply it
+    auto customZoneSets = FancyZonesDataInstance().GetCustomZoneSetsMap();
+
+    if (!customZoneSets.contains(uuid))
+    {
+        return;
+    }
+
+    FancyZonesDataTypes::ZoneSetData data{ .uuid = uuid, .type = FancyZonesDataTypes::ZoneSetLayoutType::Custom };
+    FancyZonesDataInstance().SetActiveZoneSet(workArea->UniqueId(), data);
+    FancyZonesDataInstance().SaveZoneSettings();
+    UpdateZoneSets();
+
+    if (m_settings->GetSettings()->flashZonesOnQuickSwitch)

```

Comment:
I suggest to move this out of this method and make it a stand alone function.

Reply:
That was the original plan but that function will also require a lock, which is already held. And I don't want to release the lock between this and the call to FlashZones(). Since it's just three lines I left it here.

Reply:
What about having the caller of `ApplyQuickLayout` taking the lock?
We should try to decouple the code as much as possible.

Reply:
After adding `require_write_lock` in various places it was possible to make it a standalone function, and so I did it.